keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,585 @@
+the controller processes all the Pods. The number of Pods is caped by:
+- time: in each iteration, the job controller removes all the Pods' finalizers
+  it can in a unit of time in the order of tens of seconds. This allows to
+  throttle the number of Job status updates.
+- count: Preventing big writes to etcd. We limit the number of UIDs to the order
+  of hundreds, keeping the size of the slice under 20kb.
+
+If any Pod finalizer removal fails in step 3, the controller manager still
+executes step 4 with the Pods that succeeded.
+
+Steps 2 to 4 might be skipped in the scenario where a status update happened
+too recently and the number of uncounted Pods is a small percentage of
+parallelism.
+
+### Deleted Pods
+   
+In the case where a user or another controller removes a Pod, which sets a
+deletion timestamp, the Job controller treats it the same as any other Pod.
+That is, once it reaches Failed status, the controller accounts for the Pod and
+then removes the finalizer.
+
+This is different from the legacy tracking, where the Job controller does not
+account for deleted Pods. This is a limitation that this KEP also wants to
+solve.
+
+However, if the Job controller deletes the Pod (when parallelism is decreased,
+for example), the controller removes the finalizer before deleting it. Thus,
+these deletions don't count towards the failures.
+   
+### Pod adoption
+
+If a Job with `.status.uncountedTerminatedPods != nil` can adopt a Pod
+(according to the existing adoption criteria), this Pod might not have a
+finalizer.
+
+The job controller adds the finalizer in the same patch request that modifies
+the owner reference.
+
+### Test Plan
+
+- Unit tests:
+  - Job sync with feature gate enabled.
+  - Removal of finalizers when feature gate is disabled.
+  - Tracking of terminating Pods.
+- Integration tests:
+  - Job tracking with feature enabled.
+  - Tracking of terminating Pods.
+  - Transition from feature enabled to disabled and enabled again.
+  - Tracking Jobs with big number of Pods, making sure the status is eventually
+    consistent.
+- E2E test:
+  - Job tracking with feature enabled.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implementation:
+  - Job tracking without lingering Pods
+  - Removal of finalizer when feature gate is disabled.
+- Tests: unit, integration, E2E
+
+#### Alpha -> Beta Graduation
+
+- Support for [Indexed Jobs](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+- Processing 5000 Pods per minute across any number of Jobs, with Pod creation
+  having higher priority than status updates. This might depend on
+  [Priority and Fairness](https://git.k8s.io/enhancements/keps/sig-api-machinery/1040-priority-and-fairness).
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions per Job processed within an order of
```

Comment:
If we assume that "order of minutes" is 5 minutes - that would mean O(300) QPS to delete finalizers.

I don't think we should be that aggressive - it's much more than we support now.

Reply:
I was thinking more like 10min. Is that still too much? This would be assuming that this is the only Job though.

Reply:
IMO in the situation where you have a job that's that big, 300 QPS is perfectly reasonable-- that's probably a large fraction of what you're running on the cluster anyway. Compare it to the writes it took to actually get all those pods through their lifecycle. It's not adding much.

Reply:
If you're saying how it should be in ideal world - sure, I believe even more than should be possible.
If you're asking where we are now - I doubt we will be able to do that (I mean, we can handle 300 qps of course, but we can't blindly bump qps limits until we have P&F supporting things like LISTS, and also some rate-limiting that would rate-limit watches - whatever that means).

Also note - that it highly depends on how long a particular pod takes to run - if those are very short-running, then it matters, if they are long-running, then it's not that

Reply:
This is still in the Beta -> GA graduation, so I think it should be fine to have this as tentative goal